File: test/util_caching.cpp
// google test docs
// wiki page: https://code.google.com/p/googletest/w/list
// primer: https://code.google.com/p/googletest/wiki/V1_7_Primer
// FAQ: https://code.google.com/p/googletest/wiki/FAQ
// advanced guide: https://code.google.com/p/googletest/wiki/V1_7_AdvancedGuide
// samples: https://code.google.com/p/googletest/wiki/V1_7_Samples
//
// List of some basic tests fuctions:
// Fatal assertion                      Nonfatal assertion Verifies /
// Description
//-------------------------------------------------------------------------------------------------------------------------------------------------------
// ASSERT_EQ(expected, actual);         EXPECT_EQ(expected, actual); expected ==
// actual ASSERT_NE(val1, val2);               EXPECT_NE(val1, val2); val1 !=
// val2 ASSERT_LT(val1, val2);               EXPECT_LT(val1, val2); val1 < val2
// ASSERT_LE(val1, val2);               EXPECT_LE(val1, val2); val1 <= val2
// ASSERT_GT(val1, val2);               EXPECT_GT(val1, val2); val1 > val2
// ASSERT_GE(val1, val2);               EXPECT_GE(val1, val2); val1 >= val2
//
// ASSERT_FLOAT_EQ(expected, actual);   EXPECT_FLOAT_EQ(expected, actual);   the
// two float values are almost equal (4 ULPs) ASSERT_DOUBLE_EQ(expected,
// actual); EXPECT_DOUBLE_EQ(expected, actual);  the two double values are
// almost equal (4 ULPs) ASSERT_NEAR(val1, val2, abs_error);  EXPECT_NEAR(val1,
// val2, abs_error); the difference between val1 and val2 doesn't exceed the
// given absolute error
//
// Note: more information about ULPs can be found here:
// http://www.cygnus-software.com/papers/comparingfloats/comparingfloats.htm
//
// Example of two unit test:
// TEST(Math, Add) {
//    ASSERT_EQ(10, 5+ 5);
//}
//
// TEST(Math, Float) {
//	  ASSERT_FLOAT_EQ((10.0f + 2.0f) * 3.0f, 10.0f * 3.0f + 2.0f * 3.0f)
//}
//=======================================================================================================================================================
#include <thread>
#include "gtest/gtest.h"

#include "cache.hpp"

using namespace util_caching;
using namespace std::chrono_literals;

struct SomePolicyWithoutParams {
    SomePolicyWithoutParams() = default;
    bool operator()(const Time& /*lhs*/, const Time& /*rhs*/) const {
        return true;
    }
};


class CacheTest : public ::testing::Test {
protected:
    double key1{1.};
    double key2{1.2};
    double key3{1.6};
    Time time1 = std::chrono::steady_clock::now();
    Time time2 = time1 + 10ms;
    Time time3 = time1 + 1100ms;
    Time time4 = time1 + 2100ms;
    Cache<double, double> cacheByNumber;
    Cache<Time, double> cacheByTime;
    util_caching::policies::ApproximateNumber<double> approximateNumberPolicy{0.5};
    util_caching::policies::ApproximateTime<Time, std::chrono::milliseconds> approximateTimePolicy{100};
    util_caching::policies::ApproximateTime<Time, std::chrono::seconds> approximateTimePolicy2{1};
    SomePolicyWithoutParams dummyPolicy;
};

TEST_F(CacheTest, TestWithNumberKey) {
    EXPECT_FALSE(cacheByNumber.cached(key1));
    cacheByNumber.cache(key1, 1.);

    // exact match
    EXPECT_TRUE(cacheByNumber.cached(key1));
    EXPECT_DOUBLE_EQ(*cacheByNumber.cached(key1), 1.);

    // approximate match
    EXPECT_TRUE(cacheByNumber.cached(key2, approximateNumberPolicy));
    EXPECT_DOUBLE_EQ(*cacheByNumber.cached(key2, approximateNumberPolicy), 1.);

    // over threshold
    EXPECT_FALSE(cacheByNumber.cached(key3, approximateNumberPolicy));
}


TEST_F(CacheTest, TestWithTimeKey) {
    EXPECT_FALSE(cacheByTime.cached(time1));
    cacheByTime.cache(time1, 1.);

    // exact match
    EXPECT_TRUE(cacheByTime.cached(time1));
    EXPECT_DOUBLE_EQ(*cacheByTime.cached(time1), 1.);

    // approximate match with miliseconds
    EXPECT_TRUE(cacheByTime.cached(time2, approximateTimePolicy));
    EXPECT_DOUBLE_EQ(*cacheByTime.cached(time2, approximateTimePolicy), 1.);

    // approximate match with seconds
    EXPECT_TRUE(cacheByTime.cached(time2, approximateTimePolicy2));
    EXPECT_DOUBLE_EQ(*cacheByTime.cached(time2, approximateTimePolicy2), 1.);

    // over threshold
    EXPECT_FALSE(cacheByTime.cached(time3, approximateTimePolicy));
    // exactly 1s after rounding to integer
    EXPECT_TRUE(cacheByTime.cached(time3, approximateTimePolicy2));
    // expect 2s after rounding to integer which is over threshold
    EXPECT_FALSE(cacheByTime.cached(time4, approximateTimePolicy2));
}

TEST_F(CacheTest, TestWithOtherComparisonPolicy) {
    cacheByTime.cache(time1, 1.);
    EXPECT_TRUE(cacheByTime.cached<SomePolicyWithoutParams>(time2));
    EXPECT_TRUE(cacheByTime.cached<SomePolicyWithoutParams>(time2, dummyPolicy));
}

File: test/util_caching.py
import os
import unittest
import time

from util_caching_py import number_based, time_based


class CacheTest(unittest.TestCase):
    def setUp(self):
        self.key1 = 1.0
        self.key2 = 1.2
        self.key3 = 1.6
        self.time1 = time.time()
        self.time2 = self.time1 + 0.01  # 10 milliseconds later
        self.time3 = self.time1 + 1.1  # 1100 milliseconds later
        self.time4 = self.time1 + 2.1  # 2100 milliseconds later
        self.cache_by_number = number_based.Cache()
        self.cache_by_time = time_based.Cache()
        self.approximate_number_policy = number_based.ApproximateNumber(0.5)
        self.approximate_time_policy = time_based.ApproximateTime(100)
        self.approximate_time_policy_2 = time_based.ApproximateTime(1000)

    def test_with_number_key(self):
        self.assertIsNone(self.cache_by_number.cached(self.key1))
        self.cache_by_number.cache(self.key1, 1.0)

        # exact match
        self.assertTrue(self.cache_by_number.cached(self.key1))
        self.assertEqual(self.cache_by_number.cached(self.key1), 1.0)

        # approximate match
        self.assertTrue(
            self.cache_by_number.cached(self.key2, self.approximate_number_policy)
        )
        self.assertEqual(
            self.cache_by_number.cached(self.key2, self.approximate_number_policy),
            1.0,
        )

        # over threshold
        self.assertIsNone(
            self.cache_by_number.cached(self.key3, self.approximate_number_policy)
        )

    def test_with_time_key(self):
        self.assertIsNone(self.cache_by_time.cached(self.time1))
        self.cache_by_time.cache(self.time1, 1.0)

        # exact match
        self.assertTrue(self.cache_by_time.cached(self.time1))
        self.assertEqual(self.cache_by_time.cached(self.time1), 1.0)

        # approximate match with milliseconds
        self.assertTrue(
            self.cache_by_time.cached(self.time2, self.approximate_time_policy)
        )
        self.assertEqual(
            self.cache_by_time.cached(self.time2, self.approximate_time_policy), 1.0
        )

        # approximate match with seconds
        self.assertTrue(
            self.cache_by_time.cached(self.time2, self.approximate_time_policy_2)
        )
        self.assertEqual(
            self.cache_by_time.cached(self.time2, self.approximate_time_policy_2), 1.0
        )

        # over threshold
        self.assertIsNone(
            self.cache_by_time.cached(self.time3, self.approximate_time_policy)
        )
        # expect 2s after rounding to integer which is over threshold
        self.assertIsNone(
            self.cache_by_time.cached(self.time4, self.approximate_time_policy_2)
        )


if __name__ == "__main__":
    header = "Running " + os.path.basename(__file__)

    print("=" * len(header))
    print(header)
    print("=" * len(header) + "\n")
    unittest.main(exit=False)
    print("=" * len(header) + "\n")
